jupyterhub/app.py

```diff
@@ -2884,42 +2888,46 @@ async def update_last_activity(self):
         users_count = 0
         active_users_count = 0
         now = datetime.utcnow()
-        for prefix, route in routes.items():
```

Comment:
I wonder if the two here can be more easily wrapped in a context manager/decorator, rather than large try/except blocks:

```python
def catch_db_error(f):
    """Catch and rollback database errors"""
    @functools.wraps(f)
    async def catching(self, *args, **kwargs):
        try:
            r = f(self, *args, **kwargs)
            if inspect.isawaitable(r):
                r = await r
        except SQLAlchemyError:
             self.log.exception("Rolling back session due to database error")
             self.db.rollback()
        else:
            return r

    return catching

...

@catch_db_error
async def update_last_activity(...)
```

Reply:
Thanks for the suggestion @minrk ! We made this change, can you please take a look? There is also one existing try/except in `update_last_activity `method, and we left that one unchanged.